src/main/java/org/truffleruby/language/loader/RequireNode.java

```diff
@@ -77,37 +77,60 @@ private boolean requireWithMetrics(String feature, String expandedPathRaw, Dynam
-                RubyLanguage.LOGGER
-                        .info(() -> String.format(
-                                "%s: requiring %s which is registered as an autoload for %s",
-                                RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
-                                feature,
-                                info));
-            }
+        try {
+            if (autoloadConstants != null) {
+                if (getContext().getOptions().LOG_AUTOLOAD) {
+                    String info = autoloadConstants
+                            .stream()
+                            .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
+                            .collect(Collectors.joining(" and "));
+                    RubyLanguage.LOGGER
+                            .info(() -> String.format(
+                                    "%s: requiring %s which is registered as an autoload for %s",
+                                    RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
+                                    feature,
+                                    info));
+                }
 
-            for (RubyConstant autoloadConstant : autoloadConstants) {
-                GetConstantNode.autoloadConstantStart(autoloadConstant);
-            }
-            try {
-                return doRequire(feature, expandedPath, pathString);
-            } finally {
                 for (RubyConstant autoloadConstant : autoloadConstants) {
-                    GetConstantNode.autoloadUndefineConstantIfStillAutoload(autoloadConstant);
-                    GetConstantNode.autoloadConstantStop(autoloadConstant);
-                    featureLoader.removeAutoload(autoloadConstant);
+                    GetConstantNode.autoloadConstantStart(autoloadConstant);
                 }
             }
-        } else {
+
             return doRequire(feature, expandedPath, pathString);
+        } finally {
+            final List<RubyConstant> releasedConstants = featureLoader.getAutoloadConstants(expandedPath);
+            if (releasedConstants != null) {
+                if (getContext().getOptions().LOG_AUTOLOAD) {
+                    String info = autoloadConstants
+                            .stream()
+                            .filter(c -> c.getAutoloadConstant().isAutoloading())
+                            .map(c -> c + " with " + c.getAutoloadConstant().getAutoloadPath())
+                            .collect(Collectors.joining(" and "));
+                    RubyLanguage.LOGGER
+                            .info(() -> String.format(
+                                    "%s: during requiring %s was successfully autoloaded %s",
+                                    RubyContext.fileLine(getContext().getCallStack().getTopMostUserSourceSection()),
+                                    feature,
+                                    info));
```

Comment:
It could be interesting to log which constants got loaded successfully, and which were not and were therefore undefined (if `undefineConstantIfStillAutoload` returns true, which we need to pass as the return value of `autoloadUndefineConstantIfStillAutoload`).